File: helper-apk/java/io/github/suclash/control/ProxyTileService.java
package io.github.suclash.control;

import android.service.quicksettings.Tile;
import android.service.quicksettings.TileService;

/** 控制中心磁贴：点按切换启停。 */
public class ProxyTileService extends TileService {

    @Override
    public void onStartListening() {
        super.onStartListening();
        refreshAsync();
    }

    @Override
    public void onClick() {
        refresh(TileState.TILE_STARTING);
        Root.ctlAsync("toggle", r -> {
            refreshAsync();
            // 状态文件写入可能有延迟，稍后再刷一次
            android.os.Handler h = new android.os.Handler(getMainLooper());
            h.postDelayed(this::refreshAsync, 1500);
        });
    }

    private void refreshAsync() {
        Root.execAsync("cat /data/adb/suclash/state/tile 2>/dev/null || echo off", r ->
                refresh(TileState.normalize(r.out)));
    }

    private void refresh(String state) {
        Tile t = getQsTile();
        if (t == null) return;
        boolean on = TileState.TILE_ON.equals(state) || TileState.TILE_STARTING.equals(state);
        t.setState(on ? Tile.STATE_ACTIVE : Tile.STATE_INACTIVE);
        if (TileState.TILE_STARTING.equals(state)) t.setSubtitle("启动中…");
        else if (TileState.TILE_PANIC.equals(state)) t.setSubtitle("已熔断");
        else t.setSubtitle(null);
        t.updateTile();
    }
}

File: helper-apk/src/io/github/suclash/control/ProxyTileService.kt
package io.github.suclash.control

import android.os.Build
import android.service.quicksettings.Tile
import android.service.quicksettings.TileService
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

/** 控制中心磁贴：点按切换启停。 */
class ProxyTileService : TileService() {

    override fun onStartListening() {
        super.onStartListening()
        refreshAsync()
    }

    override fun onClick() {
        refresh(TileState.TILE_STARTING)
        Root.ctlAsync("toggle") {
            refreshAsync()
            // 状态文件写入可能有延迟，稍后再刷一次
            appScope.launch(mainDispatcher) {
                delay(1_500)
                refreshAsync()
            }
        }
    }

    private fun refreshAsync() {
        Root.execAsync("cat /data/adb/suclash/state/tile 2>/dev/null || echo off") { r ->
            refresh(TileState.normalize(r.out))
        }
    }

    private fun refresh(state: String) {
        val t = qsTile ?: return
        val on = state == TileState.TILE_ON || state == TileState.TILE_STARTING
        t.state = if (on) Tile.STATE_ACTIVE else Tile.STATE_INACTIVE
        if (Build.VERSION.SDK_INT >= 31) {
            t.subtitle = when (state) {
                TileState.TILE_STARTING -> "启动中…"
                TileState.TILE_PANIC -> "已熔断"
                else -> null
            }
        }
        t.updateTile()
    }
}
